### ISSUE TITLE
Fast vjp form

### ISSUE THREAD

**Issue by ChrisRackauckas** [SciML/DiffEqFlux.jl#681]
https://github.com/SciML/DiffEqFlux.jl/pull/671 improves vjp calculations, but we could do better. The issue is that adjoints will naturally defer calculation because the memory may be used sometime in the future. But for vjps, "the future is now" in some sense.

So what we need to do is instead create a `vjp` function on `FastChain` itself which can handle the composed forward + reverse of a whole `FastDense` which is specialized when only `FastDense` exist (maybe a few other types need to be created, like a `BroadcastFunction` etc.) that are then chained into one big non-allocating pre-cached overload.

@chriselrod might have some ideas. 

@Abhishek-1Bhatt

**Comment by ChrisRackauckas** [comment]
Yes exactly. It's not always immediately called, so it allocates the variables of the forward pass into the closure to then call the future sometime in the future. But for a vjp calculation you know that "future" is right after the forward pass, so we can manually fuse the call into a single function and not allocate.

**Comment by Abhishek-1Bhatt** [comment]
We will still be using this functor for forward computation?

https://github.com/SciML/DiffEqFlux.jl/blob/cff4c35555c9d7f24803a43ae28bd56b445fceaa/src/fast_layers.jl#L76
